File: Dynamic_Programming/64.Minimum_Path_Sum/C++/64.Minimum_Path_Sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>> & grid)
    {
        //============================================//
        // Input type:                                //
        //   - vector<vector<int>> ref. (grid vector) //
        // Return type:                               //
        //   - int (result summary value)             //
        //============================================//

        /*Initialize*/
        ///// Length of rows, columns /////
        int len_rows = grid.size(), len_cols = (grid[0]).size();

        ///// Record dynamic-programming (i.e. DP) vector /////
        vector<vector<long long>> record_dp(len_rows, vector<long long>(len_cols, 0));


        /*Dynamic-programming based loop traversal*/
        (record_dp[0])[0] = (grid[0])[0]; //Update, Overwrite

        for (int cols_idx = 1; (cols_idx < len_cols); (cols_idx++)) //First rows
        {
            (record_dp[0])[cols_idx] = ((grid[0])[cols_idx] + (record_dp[0])[(cols_idx - 1)]); //Keep updating, overwriting

        } //First rows

        for (int rows_idx = 1; (rows_idx < len_rows); (rows_idx++)) //Fisrt columns
        {
            (record_dp[rows_idx])[0] = ((grid[rows_idx])[0] + (record_dp[(rows_idx - 1)])[0]); //Keep updating, overwriting

        } //Fisrt columns

        for (int rows_idx = 1; (rows_idx < len_rows); (rows_idx++)) //Rows 
        {
            for (int cols_idx = 1; (cols_idx < len_cols); (cols_idx++)) //Columns
            {
                (record_dp[rows_idx])[cols_idx] = ((grid[rows_idx])[cols_idx] + min((record_dp[(rows_idx - 1)])[cols_idx], (record_dp[rows_idx])[(cols_idx - 1)])); //Keep updating, overwriting

            } //Columns

        } //Rows

        return (int)((record_dp[(len_rows - 1)])[(len_cols - 1)]);
    }
};

File: Dynamic_Programming/64.Minimum_Path_Sum/Python/64.Minimum_Path_Sum.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        #======================================================#
        # Dynamic-programming (i.e. DP) based traversal method #
        #======================================================#

        ############
        #Initialize
        ##### Length of rows, columns #####
        len_rows, len_cols = len(grid), len(grid[0])

        ##### Record dynamic-programming (i.e. DP) array #####
        record_dp = [[0 for _ in range(len_cols)] for _ in range(len_rows)]

        
        ####################################################
        #Dynamic-programming (i.e. DP) based loop traversal
        (record_dp[0])[0] = (grid[0])[0] #Update/Overwrite

        ##### Step 1: Overwrite top-first rows, columns to matched values #####
        for cols_idx in range(1, len_cols):
            (record_dp[0])[cols_idx] = ((grid[0])[cols_idx] + (record_dp[0])[(cols_idx - 1)]) #Keep updating/overwriting

        for rows_idx in range(1, len_rows):
            (record_dp[rows_idx])[0] = ((grid[rows_idx])[0] + (record_dp[(rows_idx - 1)])[0]) #Keep updating/overwriting

        ##### Step 2: Looped-traversal with recorded array #####
        for rows_idx in range(1, len_rows):
            for cols_idx in range(1, len_cols):
                (record_dp[rows_idx])[cols_idx] = ((grid[rows_idx])[cols_idx] + min((record_dp[(rows_idx - 1)])[cols_idx], (record_dp[rows_idx])[(cols_idx - 1)])) #Keep updating/overwriting

        return (record_dp[(len_rows - 1)])[(len_cols - 1)]
